distribution/pull_v2.go

```diff
@@ -501,6 +519,29 @@ func (p *v2Puller) pullSchema2(ctx context.Context, ref reference.Named, mfst *s
 		return imageID, manifestDigest, nil
 	}
 
+	var descriptors []xfer.DownloadDescriptor
+
+	// Note that the order of this loop is in the direction of bottom-most
+	// to top-most, so that the downloads slice gets ordered correctly.
+	for _, d := range mfst.Layers {
+		layerDescriptor := &v2LayerDescriptor{
+			digest:            d.Digest,
+			repo:              p.repo,
+			repoInfo:          p.repoInfo,
+			V2MetadataService: p.V2MetadataService,
+		}
+
+		if d.MediaType == schema2.MediaTypeForeignLayer && len(d.URLs) > 0 {
+			if runtime.GOOS != "windows" {
```

Comment:
@jstarks I think we can limit the ability to windows (by blanking out urls on non-foreign, descriptors) without introducing the concept of foreign source this deeply into the code. I'm worried this abstraction is leaking too widely.